File: 0020-valid-parentheses/0020-valid-parentheses.py
"""
Time = O(n)
Space = O(n)
"""
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        brackets = {")": "(", "]": "[", "}": "{"}
        # Traverse to all brackets
        for i in range(len(s)):
            if s[i] in "([{":
                stack.append(s[i])
            else:
                if stack:
                    top = stack[-1]
                    # Check if they are matched
                    if top != brackets[s[i]]:
                        return False
                    else:
                        stack.pop()
                else:
                    return False
        
        return len(stack) == 0
            

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
#include <stack>

class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> parenMap = {{')', '('}, {']', '['}, {'}', '{'}};
        stack<char> st;

        for (char c: s) {
            if (c == '(' || c == '[' || c == '{') {
                st.push(c);
            } else {
                if (st.empty() || parenMap[c] != st.top()) {
                    return false;
                } else {
                    st.pop();
                }
            }
        }

        return st.size() == 0;
    }
};